lang/java/ipc/src/test/java/org/apache/avro/message/TestBigDecimalEvolution.java

```diff
@@ -0,0 +1,141 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class TestBigDecimalEvolution {
+
+  public static final double DELTA = 0.00000001;
+  private final MockTransceiver transceiver = new MockTransceiver();
+
+  @Test
+  public void bigDecimalEvolutionThroughLogicalType() throws Exception {
+    ReflectData oldApp = new ReflectData();
+    oldApp.addLogicalTypeConversion(new BigDecimalConversionOld());
+    ReflectData upgradeApp = new ReflectData();
+    upgradeApp.addLogicalTypeConversion(new BigDecimalConversionUpgrade());
+
+    verify(oldApp, oldApp);
+    verify(upgradeApp, upgradeApp);
+    verify(upgradeApp, oldApp);
+    verify(oldApp, upgradeApp);
+  }
+
+  private void verify(ReflectData clientData, ReflectData serverData) throws IOException {
+    Protocol clientProtocol = serverData.getProtocol(BigDecimalProtocol.class);
+    ReflectRequestor clientRequestor = new ReflectRequestor(clientProtocol, transceiver, serverData);
+    ReflectResponder clientResponder = new ReflectResponder(clientProtocol, null, serverData);
+    Schema clientRequestSchema = serverData.getSchema(BDRequest.class);
+    Schema clientResponseSchema = serverData.getSchema(BDResponse.class);
+
+    Protocol serverProtocol = clientData.getProtocol(BigDecimalProtocol.class);
+    ReflectRequestor serverRequestor = new ReflectRequestor(serverProtocol, transceiver, clientData);
+    ReflectResponder serverResponder = new ReflectResponder(serverProtocol, null, clientData);
+    Schema serverRequestSchema = clientData.getSchema(BDRequest.class);
+    Schema serverResponseSchema = clientData.getSchema(BDResponse.class);
+
+    BDRequest originalRequest = new BDRequest();
+    originalRequest.setValue(new BigDecimal("123.321"));
+
+    BDRequest request = verifyRequest(originalRequest, serverRequestSchema, clientRequestSchema, serverProtocol,
+        serverRequestor, clientResponder);
+
+    BDResponse originalResponse = new BDResponse(request.getValue());
+    verifyResponse(originalResponse, serverResponseSchema, clientResponseSchema, serverResponder, clientRequestor);
+  }
+
+  private BDRequest verifyRequest(BDRequest originalRequest, Schema writerSchema, Schema readerSchema,
+      Protocol writerProtocol, ReflectRequestor requestor, ReflectResponder responder) throws IOException {
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
+    BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(out, null);
+
+    requestor.writeRequest(writerProtocol.getMessages().get("test").getRequest(), new Object[] { originalRequest },
+        encoder);
+
+    encoder.flush();
+
+    Object obj = responder.readRequest(writerSchema, readerSchema,
+        DecoderFactory.get().binaryDecoder(out.toByteArray(), null));
+    assertTrue(obj instanceof BDRequest);
+    BDRequest request = (BDRequest) obj;
+
+    assertEquals(originalRequest.getValue().doubleValue(), request.getValue().doubleValue(), DELTA);
```

Comment:
Why use a double here? A simple equality check (optionally coupled with a "not same") is more accurate:
```
assertEquals(originalRequest.getValue(), request.getValue());
assertNotSame(originalRequest.getValue(), request.getValue());
```

Reply:
there is some issues with BigDecimal comparison. It's a kinda pain in the leg. It depends from scaling and some other underlying stuff. For example all of these asserts will fail: 
```java
assertEquals(new BigDecimal("123.321"), new BigDecimal(123.321));
assertEquals(new BigDecimal("123.321"), new BigDecimal("123.3210"));
assertEquals(new BigDecimal("123.321"), new BigDecimal("123.321").setScale(6));
```

I thought the test goal is to check the value is "the same" from global humanity POV regardless how did we transfer data.

Reply:
hmmmm... that's a valid point. Okay, I'm convinced =) 
thanks!

Reply:
done!